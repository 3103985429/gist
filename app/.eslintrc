{
  // extend root arrays
  "extends": "../.eslintrc.js",
  "parserOptions": {
    "ecmaVersion": 6,
  },
  "env": {
    "browser": false,
    "node": true
  }
}
