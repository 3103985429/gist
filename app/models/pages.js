const mongoose_client = require('./mongoose_client');
const { isNull } = require('lodash');

const mongoose = mongoose_client.mongoose;
const Schema = mongoose.Schema;

// TODO:
// the acknowledgement should be moved into an object that's connected with the image

const pageSchema = Schema({
    pageTitle: String,
    slug: String,
    metaTitle: String,
    metaDescription: String,
    rawSummary: String,
    parsedSummary: String,
    rawBody: String,
    parsedBody: String,
    rawAbout: String,
    parsedAbout: String,
    url: String,
    thumbnail: String,
    image: String,
    pdf: String,
    acknowledgement: String,
    published: { type: Boolean, default: false },
    displayOnTeamPage: { type: Boolean, default: false },
    displayOnExplorePage: { type: Boolean, default: false },
    updatedAt: { type: Date, default: Date.now },
    createdAt: { type: Date, default: Date.now },
    isArticle: Boolean,
    state: String,
    team: { type: Schema.Types.ObjectId, ref: 'Team' },
    website: { type: Schema.Types.ObjectId, ref: 'Website' },
    createdBy: { type: Schema.Types.ObjectId, ref: 'User' },
    updatedBy: { type: Schema.Types.ObjectId, ref: 'User' },
}, { timestamps: { updatedAt: 'updatedAt' } });

const deepPopulate = require('mongoose-deep-populate')(mongoose);

pageSchema.plugin(deepPopulate);

pageSchema.pre('save', function (next) {
    this.updatedAt = Date.now();
    next();
});

pageSchema.statics.buildTeamPageQuery = (user) => {
    const baseQuery = { displayOnTeamPage: true };

    if (!user) {
        return {
            ...baseQuery,
            published: true,
        };
    }

    if (user.isSuperAdmin()) {
        return baseQuery;
    }

    const userId = user._id.toString();

    return {
        ...baseQuery,
        $or: [
            {
                team: {
                    $in: user._team
                        .filter(team => team.admin.some(id => id.toString() === userId))
                        .map(team => team._id),
                },
            },
            { _id: { $in: user._articleEditors } },
            { _id: { $in: user._articleViewers } },
            { published: true },
        ],
    };
};

pageSchema.statics.getShowcasedEntries = async function () {
    const pages = await this
        .find({
            published: true,
            state: 'approved',
            displayOnExplorePage: true,
        })
        .sort('-createdAt')
        .populate([
            {
                path: 'createdBy',
                select: 'firstName lastName -_id',
            },
            {
                path: 'website',
                select: 'slug -_id',
            },
            {
                path: 'team',
                select: 'subdomain -_id',
            },
        ]);

    return pages.filter(page => !isNull(page.team));
};

const Page = mongoose.model('Page', pageSchema);

module.exports = Page;
