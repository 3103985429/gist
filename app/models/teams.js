const integerValidator = require('mongoose-integer');
const winston = require('winston');
const { isEmpty } = require('lodash');

const mongoose_client = require('./mongoose_client');

const mongoose = mongoose_client.mongoose;
const Schema = mongoose.Schema;

const team_scheme = Schema({
    title: String,
    subdomain: { type: String, unique: true },
    description: String,
    font: { type: String, default: 'centrano2' },
    logo: String,
    logo_header: String,
    externalWebsite: String,
    colorPalette: {
        type: Array,
        default: [
            '#1B6DFC',
            '#0CB7FA',
            '#FA79FC',
            '#FB3705',
            '#F9D307',
            '#FA9007',
            '#7D5807',
            '#B2E606',
            '#069908',
            '#7B07FD',
        ],
    },
    superTeam: Boolean,
    admin: [{ type: Schema.Types.ObjectId, ref: 'User' }],
    datasourceDescriptions: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    pages: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    sites: [{ type: Schema.Types.ObjectId, ref: 'Website' }],

    isEnterprise: Boolean, //if true, means it has custom view - TODO consolidate with hasEnterpriseLicense
    customViewModal: Boolean, //if true, custom view has custom view settings modal - (rhodiumgroup)
    hasEnterpriseLicense: { type: Boolean, default: false },

    google_analytics: String,
    header_script: String,
    footer_script: String,
    brandColor: {
        label: { type: String, default: '#2A2A2A' },
        select: { type: String, default: '#2A2A2A' },
    },

}, { timestamps: true });

const modelName = 'Team';

team_scheme.plugin(integerValidator);

const deepPopulate = require('mongoose-deep-populate')(mongoose);

team_scheme.plugin(deepPopulate, { whitelist: ['datasourceDescriptions.author', 'datasourceDescriptions.updatedBy', 'datasourceDescriptions.createdAt'] });

team_scheme.pre('save', function (next) {
    this._wasNew = this.isNew;
    next();
});

team_scheme.methods.notifyNewTeamCreation = function () {
    if (this._wasNew) {
        this.populate('admin', function (err, docPopulatedWithAdmin) {
            if (err || isEmpty(docPopulatedWithAdmin.admin)) {
                winston.error('Team created with error err:' + err);
            }
        });
    }
};

team_scheme.statics.getTeamEntries = async (subdomain, user, vizName) => {
    const baseQuery = { subdomain };

    if (process.env.NODE_ENV === 'enterprise' && !process.env.SUBTEAMS) {
        baseQuery.subdomain = process.env.SUBDOMAIN;
    }

    const populateQuery = [
        {
            path: 'datasourceDescriptions',
            match: require('./descriptions').buildTeamPageQuery(user, vizName),
            select: 'description uid urls title importRevision updatedAt createdAt previous_datasets updatedBy author brandColor fe_views fe_filters.default banner connection',
            populate: [
                {
                    path: 'updatedBy',
                    select: 'firstName lastName',
                },
                {
                    path: 'author',
                    select: 'firstName lastName',
                },
                {
                    path: 'previous_datasets',
                    select: 'createdAt -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
        {
            path: 'pages',
            match: require('./pages').buildTeamPageQuery(user),
            select: 'pageTitle slug parsedSummary url thumbnail image updatedBy createdBy createdAt updatedAt createdAt',
            populate: [
                {
                    path: 'createdBy',
                    select: 'firstName lastName -_id',
                },
                {
                    path: 'website',
                    select: 'slug -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
    ];

    return team.find(baseQuery).populate(populateQuery);
};

const team = mongoose.model(modelName, team_scheme);

team.GetTeams = function (fn) {
    mongoose_client.WhenMongoDBConnected(function () {
        team.find({}, function (err, teams) {
            if (err) {
                fn(err);
            } else {
                fn(null, teams);
            }
        });
    });
};

module.exports = team;
