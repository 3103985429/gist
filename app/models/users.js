const mongoose = require('mongoose');
const crypto = require('crypto');
const winston = require('winston');
const { isEmpty } = require('lodash');

const { Schema } = mongoose;

const userSchema = new Schema({
    email: {
        type: String,
        unique: true,
        required: true,
    },
    firstName: String,
    lastName: String,
    profileImageUrl: String,
    provider: String,
    hash: String,
    salt: String,
    loginAttempts: { type: Number, required: true, default: 0 },
    lockUntil: { type: Number, default: 1 },
    activated: {
        type: Boolean,
        default: false,
    },
    sampleImported: {
        type: Boolean,
        default: false,
    },
    _team: [{ type: Schema.Types.ObjectId, ref: 'Team' }],
    active: {
        type: Boolean,
        default: true,
    },
    inviteLink: String,
    // denotes those users with non-schema emails as superAdmins
    superUser: Boolean,

    invited: Object, //temp object to store invited new user
    canCreateNewViz: Array,
    canCreateNewArticle: Array,
    canCreateNewSite: Array,
    _editors: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    _viewers: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    _articleEditors: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    _articleViewers: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    _siteEditors: [{ type: Schema.Types.ObjectId, ref: 'Website' }],
    _siteViewers: [{ type: Schema.Types.ObjectId, ref: 'Website' }],
    defaultLoginTeam: { type: Schema.Types.ObjectId, ref: 'Team' },
    inviteToken: { type: Schema.Types.ObjectId, ref: 'Token' },
    confirmation_code_hash: Object,
    signedUp: {
        type: Boolean,
        default: false,
    },
    inviter: String,
    role: String,
    smartsheetToken: Object, // Smartsheet token object to hold smartsheet token connection information
    pipedriveToken: Object, // Pipedrive token object
    datadotworldToken: Object, // Data.world token object
    salesforceToken: Object, // Salesforce token object
}, { timestamps: true });

const deepPopulate = require('mongoose-deep-populate')(mongoose);
userSchema.plugin(deepPopulate, { whitelist: ['defaultLoginTeam.datasourceDescriptions', '_team.datasourceDescriptions'] });

userSchema.methods.setPassword = function (password) {
    this.salt = crypto.randomBytes(16).toString('hex');
    this.hash = crypto.pbkdf2Sync(password, this.salt, 1000, 64, 'sha512').toString('hex');
};

userSchema.methods.validPassword = function (password) {
    const hash = crypto.pbkdf2Sync(password, this.salt, 1000, 64, 'sha512').toString('hex');
    return this.hash === hash;
};

userSchema.methods.isSuperAdmin = function () {
    const re = new RegExp('\\+superuser@schemadesign.com$');
    return re.test(this.email) || this.superUser === true;
};

userSchema.methods.isAdmin = function (team) {
    if (!team) {
        winston.warn('isAdmin was called without team param');
        return false;
    }

    if (isEmpty(team.admin)) {
        winston.warn('isAdmin was called on team without admin set');
        return false;
    }

    const userId = this._id.toString();

    return team.admin.some(id => id.toString() === userId);
};

userSchema.methods.isTeamMember = function (team) {
    if (!team) {
        winston.warn('isTeamMember was called without team param');
        return false;
    }
    return this._teams.some(team => team.equals(team._id));
};

userSchema.methods.isLocked = function () {
    // check for a future lockUntil timestamp
    return this.lockUntil && this.lockUntil > Date.now();
};

userSchema.methods.checkLoginAttempts = function () {
    // Add properties to prevent brute force login if user doesn't have them
    if (!this.lockUntil) {
        this.lockUntil = 1;
    }
    if (!this.loginAttempts) {
        this.loginAttempts = 0;
    }
};

userSchema.methods.lockoutUser = function () {
    this.loginAttempts += 1;
    if (this.loginAttempts > 5) { // Lock user out after 6 failed login attempts
        this.lockUntil = Date.now() + (24 * 60 * 60 * 1000); // Lock user out for 24 hours
    }
};

userSchema.methods.resetLock = function () {
    this.loginAttempts = 0;
    this.lockUntil = 1;
};

module.exports = mongoose.model('User', userSchema);
