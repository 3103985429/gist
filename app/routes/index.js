const { defaultTo } = require('lodash');
const helmet = require('helmet');
const express = require('express');

const { errorLogger } = require('../utils/winston');
const { checkCors, API_ALLOWED } = require('./permissions');
const { version } = require('../../package.json');

const PROTOCOL = process.env.USE_SSL === 'true' ? 'https' : 'http';
const HOST = defaultTo(process.env.HOST, 'localhost:9080');
const ROOT_DOMAIN = `${PROTOCOL}://app.${HOST}`;
const API_VERSION = 'v1';

const _mountRoutes_monitoring = function (app) {
    app.get('/_ah/health', function (req, res) {
        res.set('Content-Type', 'text/plain');
        res.status(200).send('ok');
    });
};

const _mountRoutes_errorHandling = function (app) {
    // Add the error logger after all middleware and routes so that
    // it can log errors from the whole application. Any custom error
    // handlers should go after this.
    errorLogger(app);

    // eslint-disable-next-line no-unused-vars
    app.use(function (err, req, res, next) {
        const status = err.status || 404;
        const [title, message] = (status === 404) ?
            ['Oops!', 'The page you are looking for doesn\'t exist or has been moved.'] :
            ['Uh oh.', 'Something went wrong\u2026'];
        const error = process.env.NODE_ENV === 'development' ? err.toString() : `Error: ${status}`;

        res.status(status).render('error', {
            env: process.env,
            title,
            message,
            err: error,
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });

    app.use(function (req, res) {
        res.status(404);
        res.render('error', {
            env: process.env,
            title: 'Oops!',
            message: 'The page you are looking for doesn\'t exist or has been moved.',
            err: 'Error: 404',
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });
};

const _mountRoutes_endPoints = function (app) {
    app.all('*', function ({ subdomains, url }, res, next) {
        const redirectToRootDomainPatterns = [
            /^\/dashboard/,
            /^\/login/,
            /^\/signup/,
            /^\/auth/,
            /^\/product/,
            /^\/news/,
            /^\/pricing/,
            /^\/plans/,
        ];
        const isEnterprise = process.env.NODE_ENV === 'enterprise';
        const isRootDomain = subdomains.length === 1 && ['app', 'www'].includes(subdomains[0]);

        if (!isEnterprise && !isRootDomain && redirectToRootDomainPatterns.some(pattern => pattern.test(url))) {
            return res.redirect(ROOT_DOMAIN + url);
        }

        next();
    });

    app.use('/robots.txt', express.static(__dirname + '/../../robots.txt'));

    app.use('/', require('./homepage'));

    app.use('/s', require('./shared_pages'));
    app.use('/presentation', require('./stories'));
    app.use(`/json-api/${API_VERSION}`, checkCors(API_ALLOWED), function (req, res, next) {
        req.apiVersion = API_VERSION;
        next();
    }, require('./json-api/router'));
    app.use('/auth', helmet.frameguard(), checkCors(), require('./auth'));
    app.use('/login', helmet.frameguard(), checkCors(), function (req, res) {
        res.redirect('/auth/login');
    });
    app.use('/signup', helmet.frameguard(), checkCors(), require('./signup'));
    app.use('/reset', helmet.frameguard(), checkCors(), function (req, res) {
        // TODO: Perform an jwt auth here instead of in account.js / or better yet auth in both!
        res.render('auth/password', {
            env: process.env,
        });
    });

    app.use('/dashboard', helmet.frameguard(), checkCors(), require('./dashboard'));

    app.use('/api', helmet.frameguard(), require('./api'));

    app.use('/account', checkCors(), require('./account'));
    app.use('/edit-visualization', require('./external/editVisualization'));
    app.use('/create-visualization', require('./external/createVisualization'));
    app.use('/visualization-list', require('./external/visualizationList'));

    app.use('/', require('./articles'));
    app.use('/', checkCors(API_ALLOWED), require('./views').router);
};

module.exports.MountRoutes = function (app) {
    app.get('/env', function (req, res) {
        const obj = {
            node_env: process.env.NODE_ENV,
            host: HOST,
            s3Bucket: process.env.DO_S3_BUCKET,
            s3Domain: process.env.DO_S3_ENDPOINT,
            subdomain: process.env.SUBDOMAIN,
            authProtocol: process.env.AUTH_PROTOCOL,
            hasSubteams: process.env.SUBTEAMS,
            git: {
                branch: process.env.BRANCH_NAME,
                commit: process.env.COMMIT_HASH,
            },
            version,
            insight: process.env.INSIGHT,
            hide_business_links: process.env.HIDE_BUSINESS_LINKS,
            hasSmartsheetInfo: !!process.env.SMARTSHEET_CLIENT_ID,
            hasPipedriveInfo: !!process.env.PIPEDRIVE_CLIENT_ID,
            hasDatadotworldInfo: !!process.env.DATA_DOT_WORLD_CLIENT_ID,
            hasSalesforceInfo: !!process.env.SALESFORCE_CONSUMER_KEY
        };

        return res.json(obj);
    });

    _mountRoutes_monitoring(app);
    _mountRoutes_endPoints(app);
    _mountRoutes_errorHandling(app);
};
