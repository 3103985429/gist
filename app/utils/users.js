const Users = require('../models/users');

module.exports = {
    getUser,
};

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getUser(id) {
    return await Users
        .findById(id)
        .lean()
        .exec();
}
