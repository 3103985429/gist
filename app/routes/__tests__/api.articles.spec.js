import request from 'supertest';

import app from '../../../app';
import Website from '../../models/websites';
import Page from '../../models/pages';
import Team from '../../models/teams';
import User from '../../models/users';
import { hydrateDb, getId, insert, loginAndGetSessionCookie } from '../../../internals/testing/backend/utils';
import { user2 } from '../../../internals/testing/backend/fixtures/users';
import { team2 } from '../../../internals/testing/backend/fixtures/teams';
import { article1 } from '../../../internals/testing/backend/fixtures/pages';

const team2Id = getId(team2);
const user2Id = getId(user2);
const websiteData = {
    createdBy: user2Id,
    title: 'Some title',
    slug: team2.subdomain,
    team: team2Id,
    published: true,
    pages: [],
};
const pageData = {
    createdBy: user2Id,
    team: team2Id,
    slug: 'first-article',
    pageTitle: 'First Article'
};

describe('Articles API', () => {
    let appRequest;

    beforeEach(async () => {
        await hydrateDb();
        appRequest = request(app);
    });

    it('should automatically create a website if none exists', async () => {
        const url = `/api/website/search/${team2Id}/${team2.subdomain}`;
        const sessionCookie = await loginAndGetSessionCookie(appRequest, user2);
        const response = await appRequest
            .get(url)
            .set('Cookie', sessionCookie);

        expect(response).toMatchSnapshot();
    });

    it('should retrieve a website if already created', async () => {
        const website = await insert(Website, websiteData);

        const url = `/api/website/search/${team2Id}/${team2.subdomain}`;
        const sessionCookie = await loginAndGetSessionCookie(appRequest, user2);
        const response = await appRequest
            .get(url)
            .set('Cookie', sessionCookie);

        expect(response.body._id).toBe(getId(website));
        expect(response).toMatchSnapshot();
    });

    it('should create a new article (page) with the website that was previously created', async () => {
        const website = await insert(Website, websiteData);
        const requestBody = { ...pageData, website };
        const response = await appRequest.post('/api/page').send(requestBody);

        expect(response.statusCode).toBe(200);
        expect(response.body.website.pages).toContain(response.body._id);
        expect(response.body.team._id).toEqual(team2Id);

        delete response.body.team.pages;
        delete response.body.website.pages;
        response.body.createdBy._articleEditors = [];
        expect(response).toMatchSnapshot();
    });

    it('should return 500 when no website in req.body when creating a new article', async () => {
        const response = await appRequest.post('/api/page').send(pageData);

        expect(response).toMatchSnapshot();
    });

    it('should update published prop', async () => {
        const website = await insert(Website, websiteData);
        const page = await insert(Page, { ...pageData, website });
        const requestBody = { published: true };
        const response = await appRequest.put(`/api/page/${getId(page)}`).send(requestBody);

        expect(response.body.website).toBe(getId(website));

        delete response.body.website;
        expect(response).toMatchSnapshot();
    });

    it('should escape parsedBody if sent without rawBody', async () => {
        const website = await insert(Website, websiteData);
        const page = await insert(Page, { ...pageData, website });
        const requestBody = { parsedBody: "<script>alert('xss')</script>" };
        const response = await appRequest.put(`/api/page/${getId(page)}`).send(requestBody);
        delete response.body.website;

        expect(response).toMatchSnapshot();
    });

    it('should update markdown and update updatedAt field', async () => {
        const website = await insert(Website, websiteData);
        const page = await insert(Page, { ...pageData, website });
        const requestBody = {
            parsedBody: '<p>This is some content</p>\n',
            rawBody: 'This is a test\n- unordered list item 1\n1. ordered list item 1\n**strong text**\n_emphasized text_\n### heading text\n\n```\ncode text here\n```\n> quote here',
        };
        const response = await appRequest.put(`/api/page/${getId(page)}`).send(requestBody);

        expect(new Date(response.body.updatedAt).valueOf()).toBeGreaterThan(page.updatedAt.valueOf());

        delete response.body.website;
        expect(response).toMatchSnapshot();
    });

    it('should make a copy', async () => {
        const website = await insert(Website, websiteData);
        const page = await insert(Page, { ...pageData, website });
        const requestBody = {
            copyOf: getId(page),
            createdBy: user2Id,
            team: team2Id,
            makeCopy: true,
            published: false
        };
        const response = await appRequest.post('/api/page').send(requestBody);

        expect(response.body.website._id).toBe(getId(website));

        const article_copy = response.body;
        // should push that copy to the team pages array
        Team.findById(article_copy.team, (err, team) => {
            expect(team.pages).toContain(article_copy._id);
        });

        Website.findById(article_copy.website, function (err, website) {
            expect(website).not.toBeNull();
            expect(website.pages).toContain(article_copy._id);
        });

        delete response.body.team.pages;
        delete response.body.website.pages;
        expect(response).toMatchSnapshot();
    });

    it('should delete a page', async () => {
        const website = await insert(Website, websiteData);
        const { body: { _id: pageId } } = await appRequest.post('/api/page')
            .send({ ...pageData, website });
        const { body: { _id: pageCopyId } } = await appRequest.post('/api/page')
            .send({
                ...pageData,
                website,
                copyOf: pageId,
                makeCopy: true,
            });

        const response = await appRequest.delete(`/api/page/${pageCopyId}`);
        const article_copy = response.body;

        expect(response.statusCode).toBe(200);
        expect(article_copy._id).toBe(getId(website));
        expect(article_copy.pages).toContain(pageId);
        expect(article_copy.pages).not.toContain(pageCopyId);

        delete response.body.pages;
        expect(response).toMatchSnapshot();
    });

    it('should delete another article', async () => {
        const response = await appRequest.delete(`/api/page/${article1._id}`);
        expect(response.statusCode).toBe(200);

        const article_to_delete_editor_id = '5a0a007a030c6ee6bfbf3333';
        await User.findById(article_to_delete_editor_id, (err, user) => {
            expect(user._articleEditors).not.toContain(article1._id);
        });

        const article_to_delete_viewer_id = '5a0a007a030c6ee6bfbf4444';
        await User.findById(article_to_delete_viewer_id, (err, user) => {
            expect(user._articleViewers).not.toContain(article1._id);
        });

        delete response.body.pages;
        expect(response).toMatchSnapshot();
    });
});
