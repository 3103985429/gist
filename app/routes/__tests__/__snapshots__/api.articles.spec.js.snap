// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles API should automatically create a website if none exists 1`] = `
Object {
  "body": Object {
    "brandColor": Object {
      "accent": "#005CB5",
    },
    "pages": Array [],
    "published": false,
    "slug": "user",
    "team": "5a42fba8999d2a26d50fea64",
    "title": "Maitland McConnell",
  },
  "statusCode": 200,
}
`;

exports[`Articles API should create a new article (page) with the website that was previously created 1`] = `
Object {
  "body": Object {
    "createdBy": Object {
      "_articleEditors": Array [],
      "_articleViewers": Array [],
      "_editors": Array [],
      "_siteEditors": Array [],
      "_siteViewers": Array [],
      "_team": Array [
        "5a42fba8999d2a26d50fea64",
      ],
      "_viewers": Array [],
      "activated": true,
      "active": true,
      "canCreateNewArticle": Array [],
      "canCreateNewSite": Array [],
      "canCreateNewViz": Array [],
      "datadotworldToken": Object {
        "access_token": "reaaaLY_SAfE_TOOOKEN",
        "expires_at": 1e+25,
      },
      "defaultLoginTeam": "5a42fba8999d2a26d50fea64",
      "email": "user+user2@schemadesign.com",
      "firstName": "Maitland",
      "hash": "dc449e96c930ca46653ec867c448719fab39dd85c4b91c1643d2faad140f24e711da13133e34ee4da0d3a5a2ca154edd4bfaeb9cf80cb9af57bce4d7ea1b94ce",
      "lastName": "McConnell",
      "lockUntil": 1,
      "loginAttempts": 0,
      "pipedriveToken": Object {
        "access_token": "exTRa_SEeeciT_token",
        "expires_at": 1e+25,
      },
      "profileImageUrl": "https://lh3.googleusercontent.com/-XdUIqdMkCWA/AAAAAAAAAAI/AAAAAAAAAAA/4252rscbv5M/photo.jpg?sz=50",
      "provider": "local",
      "salesforceToken": Object {
        "access_token": "simple_token",
        "instance_url": "http://instance_url",
      },
      "salt": "test",
      "sampleImported": false,
      "signedUp": false,
      "smartsheetToken": Object {
        "access_token": "sUpEr_sECriT_ToKEN",
        "expires_at": 1e+25,
      },
    },
    "displayOnExplorePage": false,
    "displayOnTeamPage": false,
    "pageTitle": "First Article",
    "published": false,
    "slug": "first-article",
    "team": Object {
      "admin": Array [
        "5a42fb9e29232d26d47136dc",
      ],
      "brandColor": Object {
        "label": "#2A2A2A",
        "select": "#2A2A2A",
      },
      "colorPalette": Array [
        "#1B6DFC",
        "#0CB7FA",
        "#FA79FC",
        "#FB3705",
        "#F9D307",
        "#FA9007",
        "#7D5807",
        "#B2E606",
        "#069908",
        "#7B07FD",
      ],
      "datasourceDescriptions": Array [
        "5a42fd33999d2a26d50fea65",
        "5a42fdca29232d26d47136dd",
        "5a42fe1629232d26d4713775",
        "5b07033050f5c886278b6d2d",
      ],
      "font": "calibre",
      "hasEnterpriseLicense": false,
      "sites": Array [],
      "subdomain": "user",
      "superTeam": true,
      "title": "Maitland McConnell",
    },
    "website": Object {
      "brandColor": Object {
        "accent": "#005CB5",
      },
      "createdBy": "5a42fb9e29232d26d47136dc",
      "published": true,
      "slug": "user",
      "team": "5a42fba8999d2a26d50fea64",
      "title": "Some title",
    },
  },
  "statusCode": 200,
}
`;

exports[`Articles API should delete a page 1`] = `
Object {
  "body": Object {
    "brandColor": Object {
      "accent": "#005CB5",
    },
    "createdBy": "5a42fb9e29232d26d47136dc",
    "published": true,
    "slug": "user",
    "team": "5a42fba8999d2a26d50fea64",
    "title": "Some title",
  },
  "statusCode": 200,
}
`;

exports[`Articles API should delete another article 1`] = `
Object {
  "body": Object {
    "brandColor": Object {
      "accent": "#FEB600",
    },
    "createdBy": "5a42fb9e29232d26d47136dc",
    "homepage": Object {
      "createdBy": "5a42bf07d3bc0b20648290e0",
      "displayOnExplorePage": false,
      "displayOnTeamPage": true,
      "pageTitle": "homepage",
      "parsedBody": "<p>Snake head sea holly sweet pea cuckooflower anemone flower. Scarlet plume obedient plant flamingo flower primrose. Feverfew amaryllis bloom cosmos china aster. Spurge larkspur bloody crane&#39;s-bill marguerite sweet sultan. Blossom cowbane grass of parnassus, bellflower. Bearberry, singapore orchid chickweed Kansas feather loose strife tuberose. Lisianthus, rosebay willowherb round-leaved sundew peony tasmanian blue gum foxtail lily. Love lies bleeding.</p>
",
      "published": true,
      "rawBody": "Snake head sea holly sweet pea cuckooflower anemone flower. Scarlet plume obedient plant flamingo flower primrose. Feverfew amaryllis bloom cosmos china aster. Spurge larkspur bloody crane's-bill marguerite sweet sultan. Blossom cowbane grass of parnassus, bellflower. Bearberry, singapore orchid chickweed Kansas feather loose strife tuberose. Lisianthus, rosebay willowherb round-leaved sundew peony tasmanian blue gum foxtail lily. Love lies bleeding.",
      "slug": "homepage",
      "team": "5a42bf30d3bc0b20648290e1",
      "website": "5abeac00ef9ae15d1647721b",
    },
    "published": false,
    "slug": "user",
    "team": "5a42bf30d3bc0b20648290e1",
    "title": "user",
  },
  "statusCode": 200,
}
`;

exports[`Articles API should escape parsedBody if sent without rawBody 1`] = `
Object {
  "body": Object {
    "createdBy": "5a42fb9e29232d26d47136dc",
    "displayOnExplorePage": false,
    "displayOnTeamPage": false,
    "pageTitle": "First Article",
    "parsedBody": "&lt;script&gt;alert(&#x27;xss&#x27;)&lt;&#x2F;script&gt;",
    "published": false,
    "slug": "first-article",
    "team": "5a42fba8999d2a26d50fea64",
  },
  "statusCode": 200,
}
`;

exports[`Articles API should make a copy 1`] = `
Object {
  "body": Object {
    "createdBy": Object {
      "_articleEditors": Array [],
      "_articleViewers": Array [],
      "_editors": Array [],
      "_siteEditors": Array [],
      "_siteViewers": Array [],
      "_team": Array [
        "5a42fba8999d2a26d50fea64",
      ],
      "_viewers": Array [],
      "activated": true,
      "active": true,
      "canCreateNewArticle": Array [],
      "canCreateNewSite": Array [],
      "canCreateNewViz": Array [],
      "datadotworldToken": Object {
        "access_token": "reaaaLY_SAfE_TOOOKEN",
        "expires_at": 1e+25,
      },
      "defaultLoginTeam": "5a42fba8999d2a26d50fea64",
      "email": "user+user2@schemadesign.com",
      "firstName": "Maitland",
      "hash": "dc449e96c930ca46653ec867c448719fab39dd85c4b91c1643d2faad140f24e711da13133e34ee4da0d3a5a2ca154edd4bfaeb9cf80cb9af57bce4d7ea1b94ce",
      "lastName": "McConnell",
      "lockUntil": 1,
      "loginAttempts": 0,
      "pipedriveToken": Object {
        "access_token": "exTRa_SEeeciT_token",
        "expires_at": 1e+25,
      },
      "profileImageUrl": "https://lh3.googleusercontent.com/-XdUIqdMkCWA/AAAAAAAAAAI/AAAAAAAAAAA/4252rscbv5M/photo.jpg?sz=50",
      "provider": "local",
      "salesforceToken": Object {
        "access_token": "simple_token",
        "instance_url": "http://instance_url",
      },
      "salt": "test",
      "sampleImported": false,
      "signedUp": false,
      "smartsheetToken": Object {
        "access_token": "sUpEr_sECriT_ToKEN",
        "expires_at": 1e+25,
      },
    },
    "displayOnExplorePage": false,
    "displayOnTeamPage": false,
    "pageTitle": "First Article copy",
    "published": false,
    "slug": "first-article-copy",
    "team": Object {
      "admin": Array [
        "5a42fb9e29232d26d47136dc",
      ],
      "brandColor": Object {
        "label": "#2A2A2A",
        "select": "#2A2A2A",
      },
      "colorPalette": Array [
        "#1B6DFC",
        "#0CB7FA",
        "#FA79FC",
        "#FB3705",
        "#F9D307",
        "#FA9007",
        "#7D5807",
        "#B2E606",
        "#069908",
        "#7B07FD",
      ],
      "datasourceDescriptions": Array [
        "5a42fd33999d2a26d50fea65",
        "5a42fdca29232d26d47136dd",
        "5a42fe1629232d26d4713775",
        "5b07033050f5c886278b6d2d",
      ],
      "font": "calibre",
      "hasEnterpriseLicense": false,
      "sites": Array [],
      "subdomain": "user",
      "superTeam": true,
      "title": "Maitland McConnell",
    },
    "website": Object {
      "brandColor": Object {
        "accent": "#005CB5",
      },
      "createdBy": "5a42fb9e29232d26d47136dc",
      "published": true,
      "slug": "user",
      "team": "5a42fba8999d2a26d50fea64",
      "title": "Some title",
    },
  },
  "statusCode": 200,
}
`;

exports[`Articles API should retrieve a website if already created 1`] = `
Object {
  "body": Object {
    "brandColor": Object {
      "accent": "#005CB5",
    },
    "createdBy": "5a42fb9e29232d26d47136dc",
    "pages": Array [],
    "published": true,
    "slug": "user",
    "team": "5a42fba8999d2a26d50fea64",
    "title": "Some title",
  },
  "statusCode": 200,
}
`;

exports[`Articles API should return 500 when no website in req.body when creating a new article 1`] = `
Object {
  "body": Object {},
  "statusCode": 500,
}
`;

exports[`Articles API should update markdown and update updatedAt field 1`] = `
Object {
  "body": Object {
    "createdBy": "5a42fb9e29232d26d47136dc",
    "displayOnExplorePage": false,
    "displayOnTeamPage": false,
    "pageTitle": "First Article",
    "parsedBody": "<p>This is a test</p>
<ul>
<li>unordered list item 1</li>
</ul>
<ol>
<li>ordered list item 1
<strong>strong text</strong>
<em>emphasized text</em><h3 id=\\"heading-text\\">heading text</h3>
</li>
</ol>
<pre><code>code text here</code></pre><blockquote>
<p>quote here</p>
</blockquote>
",
    "published": false,
    "rawBody": "This is a test
- unordered list item 1
1. ordered list item 1
**strong text**
_emphasized text_
### heading text

\`\`\`
code text here
\`\`\`
> quote here",
    "slug": "first-article",
    "team": "5a42fba8999d2a26d50fea64",
  },
  "statusCode": 200,
}
`;

exports[`Articles API should update published prop 1`] = `
Object {
  "body": Object {
    "createdBy": "5a42fb9e29232d26d47136dc",
    "displayOnExplorePage": false,
    "displayOnTeamPage": false,
    "pageTitle": "First Article",
    "published": true,
    "slug": "first-article",
    "team": "5a42fba8999d2a26d50fea64",
  },
  "statusCode": 200,
}
`;
